PLMS.BLL/ServicesImplementation/GoalService.cs

```diff
@@ -0,0 +1,165 @@
+
+            _goalRepository.Update(goal);
+            await _unitOfWork.CommitChangesToDatabaseAsync();
+        }
+
+        public async Task DeleteGoalAsync(int id, string userId)
+        {
+            Goal? goal = await _goalRepository.GetByPredicateAsync(g => g.Id == id);
+            if (goal == null) 
+            { 
+                return; 
+            }
+            if (!await _permissionService.HasPermissionToCategory(goal.CategoryId, userId))
+            {
+                throw new NotFoundException("Goal was not found.");
+            }
+            _goalRepository.Remove(goal);
+            await _unitOfWork.CommitChangesToDatabaseAsync();
+        }
+
+        public async Task<GetGoalDto> GetGoalByIdAsync(int id, string userId)
+        {
+            var goal = await _goalRepository.GetByPredicateAsync(g => g.Id == id, g => g.Category, g => g.Status, g => g.Priority, g => g.Tasks) ?? throw new NotFoundException("Goal was not found");
+
+            if (!_permissionService.HasPermissionToCategory(goal.Category, userId))
+            {
+                throw new NotFoundException("Goal was not found.");
+            }
+
+            var goalDto = _mapper.Map<GetGoalDto>(goal);
+
+            return goalDto;
+        }
+
+        public async Task<GoalCompletionInfoDto> GetTaskCompletionPercentageAsync(string userId)
+        {
+            var taskCompletionPercentage = await _goalRepository.GetAll().Where(g => g.UserId == userId).Select(g => new 
+            GoalCompletionInfoDto.GoalInfoDto
+            {
+                GoalId = g.Id,
+                GoalTitle = g.Title,
+                CompletionPercentage = g.Tasks.Any() ? (g.Tasks.Count(t => t.StatusId == (int)StatusEnum.Completed) * 100.0) / g.Tasks.Count() : 0.0
+            }).ToListAsync();
+
+
+            return new GoalCompletionInfoDto { Goals = taskCompletionPercentage };
+        }
+
+        public async Task<List<GetGoalDto>> GetFilteredGoalsAsync(GoalFilter filters)
+        {
+            var filterExpression = ApplyFilters(filters);
+            var orderFunction = ApplyOrder(filters);
+
+            // Apply filtation and sorting on the DB level.
+            var query = _goalRepository.GetAll()
+                                       .Include(t => t.Category)
+                                       .Include(g => g.GoalComments)
+                                       .Include(t => t.Priority)
+                                       .Include(t => t.Status)
+                                       .Where(filterExpression);
+
+            var orderedQuery = orderFunction(query);
+
+            // Pagination
+            var paginatedQuery = orderedQuery.Skip((filters.PageNumber - 1) * filters.ItemsPerPageCount)
+                                             .Take(filters.ItemsPerPageCount);
+
+            // Call to the DB
+            var goals = await paginatedQuery.ToListAsync();
+
+            var goalsDto = _mapper.Map<List<GetGoalDto>>(goals);
```

Comment:
Make sure Goals and Tasks are also sorted (just standard sorting for consistency)

Reply:
fixed

Reply:
Use OrderBy or OrderByDesc instead of Sort method and make sort meaningful to the context - if you sort comments, do it by CreatedAt field, for example.
You should end up with something like this: `g.GoalComments.OrderByDescending(_ => _.CreatedAt)`
For further detais on OrderBy vs Sort read this: https://stackoverflow.com/questions/1832684/c-sharp-sort-and-orderby-comparison

Reply:
Fixed